### PR TITLE
Implement reading password from pipe/stdin

### PR DESCRIPTION
**Pull request by thalman**
New option --stdin/-s is available for root user. It is useful
for automation/setup and it makes shadow utils passwd more versatile.

### CODE REVIEW

File: src/passwd.c [hunk 1]
```diff
@@ -266,56 +268,69 @@ static int new_password (const struct passwd *pw)
...
 		if (NULL == cp) {
-			MEMZERO(orig);
-			MEMZERO(pass);
 			return -1;
 		}
-		if (warned && (strcmp (pass, cp) != 0)) {
-			warned = false;
-		}
-		STRTCPY(pass, cp);
+		STRTCPY (pass, cp);
```
Comment (alejandro-colomar) [inline src/passwd.c:279]: Hmmmm, I need to come back some day to string hardening, and check what to do when STRTCPY() doesn't fit.

From what I can see, this is silently clamping the password to the first 199 bytes +'\0'.  That doesn't look like a good thing to do.

File: src/passwd.c [hunk 2]
```diff
@@ -682,11 +697,11 @@ static void update_shadow (void)
  *	-l	lock the password of the named account (*)
  *	-n #	set sp_min to # days (*)
  *	-r #	change password in # repository
- *	-s	execute chsh command to interpret flags
```
Comment (alejandro-colomar) [inline src/passwd.c:685]: What heppened to -s?

File: src/passwd.c [hunk 3]
```diff
@@ -850,6 +866,16 @@ int main (int argc, char **argv)
 				xflg = true;
 				anyflag = true;
 				break;
+			case 't':
+				if (!amroot) {
```
Comment (alejandro-colomar) [inline src/passwd.c:870]: Ahh, okay, you have the check here.  Forget what I said above.

**Comment by stoeckmann** [comment]
> It is useful for automation/setup and it makes shadow utils passwd more versatile.

Which use case does this cover compared to what `newusers` has to offer? The advantage of `newusers` is that it's not setuid. Pretty much "root only" already.

File: src/passwd.c (hunk 1 above)
Comment (stoeckmann) [reply src/passwd.c:279]: Yes, that's what's happening: https://github.com/shadow-maint/shadow/issues/886

Comment (alejandro-colomar) [reply src/passwd.c:279]: Thanks!

**Comment by thalman** [comment]
Hmm, I found that my PR does not work correctly with PAM. Let me investigate it first.

File: src/passwd.c (hunk 2 above)
Comment (alejandro-colomar) [reply src/passwd.c:685]: Please don't remove that line.

**Review by alejandro-colomar — DISMISSED** [review]
Please don't remove the `-s` comment in this PR.

File: src/passwd.c (hunk 2 above)
Comment (thalman) [reply src/passwd.c]: It seems that -s was never there or it is long gone. I hesitated if I can usurp -s short flag for --stdin. Now another PR has been merged that cleaned this comment so I updated my PR to
- not remove old comment
- use -s as a short option for --stdin

**Comment by thalman** [comment]
> Please don't remove the `-s` comment in this PR. (Never mind; fixed in master.)

Removed from PR

File: src/passwd.c (hunk 2 above)
Comment (alejandro-colomar) [reply src/passwd.c]: Ahh, makes sense.  I checked, and none of FreeBSD, NetBSD, nor OpenBSD use -s, so that LGTM.

**Review by ikerexxe — APPROVED** [review]
LGTM. Thanks for the patch and your hard work here!

**Comment by stoeckmann** [comment]
> > Which use case does this cover compared to what `newusers` has to offer? The advantage of `newusers` is that it's not setuid. Pretty much "root only" already.
> 
> But yes you are right, this can be achieved in many different ways.

Just to add to my initial remark: The solution with `newusers` has a critical issue: Passwords cannot contain colons because they are field separators. So I'm not a fan of newusers anymore at all.

Thank you for clarification!

**Pull request merged**